cmd/otel-allocator/allocation/least_weighted/allocator.go

```diff
@@ -0,0 +1,210 @@
+// This method is called from within SetTargets and SetCollectors, whose caller
+// acquires the needed lock.
+func (allocator *LeastWeightedAllocator) findNextCollector() *strategy.Collector {
+	var col *strategy.Collector
+	for _, v := range allocator.state.Collectors() {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col == nil {
+			col = v
+		} else {
+			if v.NumTargets < col.NumTargets {
+				col = v
+			}
+		}
+	}
+	return col
+}
+
+// addTargetToTargetItems assigns a target to the next available collector and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, whose caller acquires the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+func (allocator *LeastWeightedAllocator) addTargetToTargetItems(target *strategy.TargetItem) {
+	chosenCollector := allocator.findNextCollector()
+	targetItem := &strategy.TargetItem{
+		JobName:       target.JobName,
+		Link:          strategy.LinkJSON{Link: fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL:     target.TargetURL,
+		Label:         target.Label,
+		CollectorName: chosenCollector.Name,
+	}
+	allocator.state.SetTargetItem(targetItem.Hash(), targetItem)
+	chosenCollector.NumTargets++
+	strategy.TargetsPerCollector.WithLabelValues(chosenCollector.Name).Set(float64(chosenCollector.NumTargets))
+}
+
+func (allocator *LeastWeightedAllocator) handleTargets(diff utility.Changes[*strategy.TargetItem]) {
+	// Check for removals
+	for k, target := range allocator.state.TargetItems() {
+		// if the current target is in the removals list
+		if _, ok := diff.Removals()[k]; ok {
+			c := allocator.state.Collectors()[target.CollectorName]
+			c.NumTargets--
+			allocator.state.RemoveTargetItem(k)
+			strategy.TargetsPerCollector.WithLabelValues(target.CollectorName).Set(float64(c.NumTargets))
+		}
+	}
+
+	// Check for additions
+	for k, target := range diff.Additions() {
+		// Do nothing if the item is already there
+		if _, ok := allocator.state.TargetItems()[k]; ok {
+			continue
+		} else {
+			// Assign new set of collectors with the one different name
+			allocator.addTargetToTargetItems(target)
+		}
+	}
+}
+
+func (allocator *LeastWeightedAllocator) handleCollectors(diff utility.Changes[*strategy.Collector]) {
+	// Clear existing collectors
+	for _, k := range diff.Removals() {
+		allocator.state.RemoveCollector(k.Name)
+		strategy.TargetsPerCollector.WithLabelValues(k.Name).Set(0)
+	}
+	// Insert the new collectors
+	for _, i := range diff.Additions() {
+		allocator.state.SetCollector(i.Name, &strategy.Collector{Name: i.Name, NumTargets: 0})
+	}
+
+	// find targets which need to be redistributed
+	var redistribute []*strategy.TargetItem
+	for _, item := range allocator.state.TargetItems() {
+		for _, s := range diff.Removals() {
```

Comment:
Instead of cycling through all the removed collectors, can we check if the collector is still in the collectors map? If not, the target needs to be redistributed.